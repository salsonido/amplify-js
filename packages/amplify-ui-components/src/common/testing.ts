import {
	boolean as booleanKnob,
	select as selectKnob,
	text as textKnob,
} from '@storybook/addon-knobs';

export const pixelThreshold = 10;

export const selectors = {
	// amplify-form-field
	formFieldDescription: '[data-test=form-field-description]',
};

export const knobs = {
	labelKnob(labelPlaceholder: string) {
		return textKnob('Label', labelPlaceholder);
	},

	descriptionKnob(descriptionPlaceholder: string) {
		return textKnob('Description', descriptionPlaceholder);
	},

	hintKnob(hintPlaceholder: string) {
		return textKnob('Hint', hintPlaceholder);
	},

	inputTypeKnob(inputTypePlaceholder: string) {
		return selectKnob(
			'Type',
			[
				'date',
				'email',
				'number',
				'password',
				'search',
				'tel',
				'text',
				'url',
				'time',
			],
			inputTypePlaceholder
		);
	},

	placeholderKnob(placeholderPlaceholder: string) {
		return textKnob('Placeholder', placeholderPlaceholder);
	},

	buttonTypeKnob(buttonTypePlaceholder: string) {
		return selectKnob(
			'Button type',
			['button', 'submit', 'reset'],
			buttonTypePlaceholder
		);
	},

	buttonTextKnob(buttonTextPlaceholder: string) {
		return textKnob('Button text', buttonTextPlaceholder);
	},

	toggleKnob(toggleDefault: boolean) {
		return booleanKnob('Toggle', toggleDefault);
	},

	disabledKnob(disabledDefault: boolean) {
		return booleanKnob('Disabled', disabledDefault);
	},

	formSectionHeaderKnob(headerPlaceholder: string) {
		return textKnob('Header text', headerPlaceholder);
	},

	formSectionSubmitKnob(submitPlaceholder: string) {
		return textKnob('Submit button text', submitPlaceholder);
	},

	tooltipAutoShowKnob(autoShowDefault: boolean) {
		return booleanKnob('Tooltip autoshow', autoShowDefault);
	},
};
