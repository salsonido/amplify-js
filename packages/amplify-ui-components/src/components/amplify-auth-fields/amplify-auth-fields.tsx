import { Component, Prop, h } from '@stencil/core';
import { FormFieldType, FormFieldTypes } from './amplify-auth-fields-interface';
import componentFieldMapping from './component-field-mapping';

@Component({
	tag: 'amplify-auth-fields',
	styleUrl: 'amplify-auth-fields.scss',
})
export class AmplifyAuthFields {
	/**
	 * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
	 * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
	 * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
	 * ```
	 * [
	 *  {
	 *    type: string,
	 *    label: string,
	 *    placeholder: string,
	 *    hint: string | Functional Component | null,
	 *    required: boolean
	 *  }
	 * ]
	 * ```
	 */
	@Prop() formFields: FormFieldTypes | string[];

	private constructFormFieldOptions(formFields: FormFieldTypes | string[]) {
		const content = [];

		if (formFields === undefined) return '';

		formFields.forEach((formField: FormFieldType | string) => {
			if (typeof formField === 'string') {
				content.push(componentFieldMapping[formField](formField));
			} else if (Object.keys(componentFieldMapping).includes(formField.type)) {
				content.push(componentFieldMapping[formField.type](formField));
			} else {
				content.push(componentFieldMapping['default'](formField));
			}
		});

		return content;
	}

	render() {
		return (
			<div class="auth-fields">
				{this.constructFormFieldOptions(this.formFields)}
			</div>
		);
	}
}
