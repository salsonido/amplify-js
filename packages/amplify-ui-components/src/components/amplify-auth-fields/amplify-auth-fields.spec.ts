import { newSpecPage } from '@stencil/core/testing';
import { AmplifyAuthFields } from './amplify-auth-fields';

describe('amplify-auth-fields spec:', () => {
	describe('Component logic ->', () => {
		let authFields;

		beforeEach(() => {
			authFields = new AmplifyAuthFields();
		});

		it('should return form fields as being defined when passing FormFieldTypes', () => {
			const constructFormFieldOptionsMock = jest.spyOn(
				authFields,
				'constructFormFieldOptions'
			);
			const result = [
				{ type: 'username', label: 'New username' },
				{ type: 'password', label: 'New password' },
			];

			authFields.formFields = [
				{ type: 'username', label: 'New username' },
				{ type: 'password', label: 'New password' },
			];

			authFields.constructFormFieldOptions();

			expect(authFields.formFields).toEqual(result);
			expect(constructFormFieldOptionsMock).toHaveBeenCalledTimes(1);
		});

		it('should return form fields as being defined when passing `username` and `password` as parameter', () => {
			const constructFormFieldOptionsMock = jest.spyOn(
				authFields,
				'constructFormFieldOptions'
			);
			const result = ['username', 'password'];

			authFields.formFields = ['username', 'password'];

			authFields.constructFormFieldOptions();

			expect(authFields.formFields).toEqual(result);
			expect(constructFormFieldOptionsMock).toHaveBeenCalledTimes(1);
		});

		it('should return the formFields as undefined by default', () => {
			expect(authFields.formFields).toBeUndefined();
		});
	});
	describe('Render logic ->', () => {
		it('should render with a `username` and `password` field', async () => {
			const usernameField = ['username', 'password'];

			const page = await newSpecPage({
				components: [AmplifyAuthFields],
			});

			const component = page.doc.createElement('amplify-auth-fields');
			(component as any).formFields = usernameField;
			page.body.appendChild(component);
			await page.waitForChanges();

			expect(page.root).toMatchSnapshot();
		});

		it('should render with no default values and a empty string', async () => {
			const page = await newSpecPage({
				components: [AmplifyAuthFields],
				html: `<amplify-auth-fields></amplify-auth-fields>`,
			});

			expect(page.root).toMatchSnapshot();
		});

		it('should render custom type fields', async () => {
			const customField = [{ label: 'date', type: 'date' }];

			const page = await newSpecPage({
				components: [AmplifyAuthFields],
			});

			const component = page.doc.createElement('amplify-auth-fields');
			(component as any).formFields = customField;
			page.body.appendChild(component);
			await page.waitForChanges();

			expect(page.root).toMatchSnapshot();
		});
	});
});
