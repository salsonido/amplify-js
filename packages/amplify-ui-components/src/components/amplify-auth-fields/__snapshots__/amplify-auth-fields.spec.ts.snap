// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-auth-fields spec: Render logic -> should render custom type fields 1`] = `
<amplify-auth-fields>
  <div class="auth-fields">
    <amplify-form-field label="date" type="date"></amplify-form-field>
  </div>
</amplify-auth-fields>
`;

exports[`amplify-auth-fields spec: Render logic -> should render with a \`username\` and \`password\` field 1`] = `
<amplify-auth-fields>
  <div class="auth-fields">
    <amplify-username-field></amplify-username-field>
    <amplify-password-field></amplify-password-field>
  </div>
</amplify-auth-fields>
`;

exports[`amplify-auth-fields spec: Render logic -> should render with no default values and a empty string 1`] = `
<amplify-auth-fields>
  <div class="auth-fields"></div>
</amplify-auth-fields>
`;
